package/kedro_viz/data_access/repositories/modular_pipelines.py

```diff
@@ -1,246 +1,324 @@
+
+        Args:
+            nested_namespace (str): The nested namespace to expand.
+
+        Returns:
+            List[str]: A list of expanded namespaces.
+            Example:
+                >>> nested_namespace = "uk.data_processing.internal"
+                >>> modular_pipeline_repo_obj = ModularPipelinesRepository()
+                >>> expanded_namespace =
+                        modular_pipeline_repo_obj._explode_namespace(nested_namespace)
+                >>> expanded_namespace
+                ["uk", "uk.data_processing", "uk.data_processing.internal"]
+        """
+        if not nested_namespace or "." not in nested_namespace:
+            return [nested_namespace] if nested_namespace else []
+        ns_parts = nested_namespace.split(".")
+        return [".".join(ns_parts[: i + 1]) for i in range(len(ns_parts))]
 
     def get_or_create_modular_pipeline(
         self, modular_pipeline_id: str
     ) -> ModularPipelineNode:
-        """Get the modular pipeline node with the given ID from the repository.
-        If it doesn't exist, create the node, add to the repository and return the instance.
+        """
+        Retrieve or create a modular pipeline node.
 
         Args:
-            modular_pipeline_id: The ID of the modular pipeline to retrieve from the repository.
+            modular_pipeline_id (str): The ID of the modular pipeline to retrieve or create.
+
         Returns:
-            A ModularPipelineNode instance with the given ID.
-        Example:
-            >>> modular_pipeline_repository = ModularPipelinesRepository()
-            >>> modular_pipeline_node = modular_pipeline_repository.get_or_create_modular_pipeline(
-            ...     "data_science"
-            ... )
-            >>> assert modular_pipeline_node.id == "data_science"
+            ModularPipelineNode: The retrieved or newly created modular pipeline node.
         """
+
         if not self.has_modular_pipeline(modular_pipeline_id):
-            modular_pipeline_node = GraphNode.create_modular_pipeline_node(
+            self.tree[modular_pipeline_id] = GraphNode.create_modular_pipeline_node(
                 modular_pipeline_id
             )
-            self.tree[modular_pipeline_id] = modular_pipeline_node
         return self.tree[modular_pipeline_id]
 
-    def add_input(
-        self,
-        modular_pipeline_id: str,
-        input_node: Union[DataNode, TranscodedDataNode, ParametersNode],
-    ) -> None:
-        """Add an input to a modular pipeline based on whether it's an internal or external input.
-        The input to a modular pipeline can only be a data node or parameter node.
-        The input also has knowledge of which modular pipelines it belongs to
-        based on its namespace. This information can be accessed with through
-        `input_node.modular_pipelines`.
+    def add_inputs(self, modular_pipeline_id: str, inputs: Set[str]) -> None:
+        """
+        Add input datasets to the modular pipeline.

```

Comment:
In the code there's also a special case for `parameters`. I think it makes sense to mention that in the docstring.

Reply:
Yes there is a special case for parameters and that is while adding the datasets as children. That is in the docstring for `add_children`, `_add_children_to_parent_pipeline` and `_add_datasets_as_children`. The method here is to add_inputs to the modular pipeline (like the free_inputs hanging outside the modular pipeline)

Reply:
I still think it's a good idea to describe the following code in the docstring:

```
 if is_dataset_param(_input):
                self.parameters.add(hashed_input)
```